pkg/start/start.go

```diff
@@ -115,6 +117,22 @@ func NewOptions() *Options {
 	}
 }
 
+func cvoGatesFrom(gate *configv1.FeatureGate) cvo.FeatureGates {
+	var enabledGates cvo.FeatureGates
+	for _, g := range gate.Status.FeatureGates {
+		if g.Version != version.Version.String() {
+			continue
```

Comment:
I see this approach discussed briefly in openshift/api#1405, but it feels a bit weird.  If a gate claims enabled in 4.y.z and we happen to be running a different version (e.g. because we're updating into a new CVO) it looks like we'll disble all the gates until whatever manages `FeatureGate` catches up with us.  But we don't want to blip these off early in the update, do we?  We might be unique vs. other existing consumers in going first in the update, while most other consumers trail behind the Kube API server operator (except etcd, which is in parallel).  Anyhow, I'm not entirely sure how this plays out, and it would be nice to have the commit message talk me through the behavior as we understand it :)

Reply:
Cool, I think I misunderstood how exactly the `FeatureGate` works during upgrades - I for some reason believed the resource is managed by the config-operator for some reason I lived in the world where CVO is upgraded after `config-operator`, but that does not seem to be true (`0000_00_cluster-version-operator_03_deployment` goes before `0000_10_config-operator_08_clusteroperator`. I'm clearly the ideal person to write these PRs :-P

I started following [o/enhancements dev-guide/featuresets.md](https://github.com/openshift/enhancements/blob/master/dev-guide/featuresets.md#openshiftyour-repo) that lead me to [`FeatureGateAccess`](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/featuregates/simple_featuregate_reader.go#L313) where the controller errors and requeues until the gates are set, and the dev-guide setup code makes the controller startup block until that happens. That felt scary for CVO so I thought running briefly without gates enalbed (which is fine for the current state of gates in CVO where there is one which is off by default) is acceptable (aggravated by my false belief that `featuregatestopper` would kill us once `status` finally updates).

I guess our only two real options is to assume some sane default or block? And I guess we cannot really block because of our role in the upgrades because we upgrade before config-operator so we would never see the new version gates?

You are totally correct that whatever corners we cut, we need to document them better. I'll spend more time looking into this and at the very least write much better commit messages.